src/convnets/resnets/core.jl

```diff
@@ -0,0 +1,468 @@
+    end
+end
+
+function _make_downsample_fns(sym::Symbol, layers)
+    @assert sym in keys(shortcut_dict)
+    "The shortcut type must be one of $(sort(collect(keys(shortcut_dict))))"
+    return collect(shortcut_dict[sym] for _ in 1:length(layers))
+end
+_make_downsample_fns(tup::NTuple{2}, layers) = collect(tup for _ in 1:length(layers))
+
+# Makes the main stages of the ResNet model. This is an internal function and should not be 
+# used by end-users. `block_fn` is a function that returns a single block of the ResNet. 
+# See `basicblock` and `bottleneck` for examples. A block must define a function 
+# `expansion(::typeof(block))` that returns the expansion factor of the block.
+function _make_blocks(block_fn, channels, block_repeats, inplanes; output_stride = 32,
+                      downsample_fns::Vector, drop_rates::NamedTuple,
+                      block_args::NamedTuple)
+    @assert output_stride in (8, 16, 32) "Invalid `output_stride`. Must be one of (8, 16, 32)"
+    expansion = expansion_factor(block_fn)
+    stages = []
+    net_block_idx = 1
+    net_stride = 4
+    dilation = prev_dilation = 1
+    # Stochastic depth linear decay rule (DropPath)
+    dp_rates = LinRange{Float32}(0.0, get(drop_rates, :drop_path_rate, 0),
+                                 sum(block_repeats))
+    # DropBlock rate
+    dbr = get(drop_rates, :drop_block_rate, 0)
+    ## Construct each stage
+    for (stage_idx, itr) in enumerate(zip(channels, block_repeats, _drop_blocks(dbr),
+                                          downsample_fns))
+        # Number of planes in each stage, number of blocks in each stage, and the drop block rate
+        planes, num_blocks, drop_block, down_fns = itr
+        # Stride calculations for each stage
+        stride = stage_idx == 1 ? 1 : 2
+        if net_stride >= output_stride
+            dilation *= stride
+            stride = 1
+        else
+            net_stride *= stride
+        end
+        # Downsample block; either a (default) convolution-based block or a pooling-based block
+        downsample = downsample_block(down_fns, inplanes, planes, expansion;
+                                      stride, dilation)
+        ## Construct the blocks for each stage
+        blocks = []
+        for block_idx in 1:num_blocks
+            # Different behaviour for the first block of each stage
+            downsample = block_idx == 1 ? downsample : identity
+            stride = block_idx == 1 ? stride : 1
+            push!(blocks,
+                  block_fn(inplanes, planes; stride, downsample,
+                           first_dilation = prev_dilation,
+                           drop_path = DropPath(dp_rates[block_idx]), drop_block,
+                           block_args...))
+            prev_dilation = dilation
+            inplanes = planes * expansion
+            net_block_idx += 1
+        end
+        push!(stages, Chain(blocks...))
+    end
```

Comment:
My thought was that these two loops are the core logic of `_make_blocks`. So instead of adding logic such as dilation handling, choosing whether to insert downsampling layers, etc. into `_make_blocks` itself, we could pass in a callback which takes the `stage_idx` + `block_idx` and handles the logic itself:
```julia
function _make_blocks(block_fn, channels, block_repeats, downsample_fns)
   ...
  for (stage_idx, itr) in enumerate(zip(channels, block_repeats, downsample_fns))
    for block_idx in 1:num_blocks
      block_fn(stage_idx, block_idx, itr...)
    end
  end
end
```
This is pseudo-Julia, but even with the full logic it may be short enough to consider inlining into `resnet`. Note how other arguments such as `drop_rates` or `block_args` no longer need to be threaded all the way through to `_make_blocks`. They can be handled by the callback, which may be a callable struct or closure. Callbacks which don't have special behaviour for certain stages/blocks can simply ignore those arguments, whereas ones which do (e.g. the comment `# Different behaviour for the first block of each stage` in the inner loop here can handle that accordingly.

Reply:
One could imagine extending this one layer up the stack and making `resnet` take a similar callback. The downside would be that the low-level resnet API handles less for you, but the upside is that it becomes more flexible and less threading of arguments is required. The big question I suppose is whether the "mid-level" API can make up for the slightly more involved process (my feeling is yes, but that's just a feeling).

Reply:
To be clear, are you suggesting that `drop_rates` be handled directly by the `block_fn` callback? This is the idea I had too, but then we go back to the question of restricting `block_fn` down to a minimal interface. The solution I had for this was to have something like 

```julia
block_layers = block_fn(inplanes, outplanes, stride, dilation)
block = _make_block(block_layers, block_idx, stage_idx; kwargs...)
```

Now `_make_block` can handle the `drop_rates` and other block arguments not essential (`downsample`, `attn_layer` etc).

Reply:
So these two lines are run in the loop body? That could work too, though my understanding was that you have models for which `block_fn` depends on the block and stage indices. FWIW, I am 100% in favour of a minimal interface: the grand vision (i.e. not for this PR) is that I can create a "resnet" using `log` (substitute with your trivial function of choice) and `identity`.

Reply:
> So these two lines are run in the loop body?

Yes.

> my understanding was that you have models for which `block_fn` depends on the block and stage indices

`stride` and `dilation` will not abstract out in this interface because those calculations will still unfortunately have to be done in `_make_blocks`. But I feel that's a small price to pay for a slightly tighter interface. And having those calculations in a more visible function should hopefully also help new devs understand the programming logic a little easier

Reply:
I could be mistaken, but doesn't having one fixed scheme for calculating stride/dilation preclude using this API for isotropic resnet variants?

Reply:
The other simple thing to do could be to have `block_fn` _slightly_ more wide and take in `block_idx` and `stage_idx` as well (could make it a `Tuple`). Then all the calculations abstract out. But `block_idx` and `stage_idx` won't be _necessary_ for any custom `block_fn` to work, so it's still on course for the grand vision

Reply:
That was my idea, yeah. In a sense that (and any derived parameters, like drop rate) is the only information `block_fn` can't collect for itself, because it comes from iterating over stages/block indices in `_make_blocks`.

Reply:
Yep. I'll still try and keep it two functions - this means that stuff like `drop_rates` will not live in the `block_fn` - but I think this compromise is a rather good start. Maybe future PRs can iterate on design, but I'm satisfied for now 😄 I'll start hacking on it